pkg/tfshim/sdk-v2/proposed_new.go

```diff
@@ -16,75 +16,296 @@ package sdkv2
 
 import (
 	hcty "github.com/hashicorp/go-cty/cty"
-	hctyjson "github.com/hashicorp/go-cty/cty/json"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
 	"github.com/zclconf/go-cty/cty"
-	ctyjson "github.com/zclconf/go-cty/cty/json"
 
 	"github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfshim/sdk-v2/internal/tf/configs/configschema"
 	"github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfshim/sdk-v2/internal/tf/plans/objchange"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/contract"
 )
 
 func proposedNew(res *schema.Resource, prior, config hcty.Value) (hcty.Value, error) {
 	schema, err := configschemaBlock(res)
 	if err != nil {
 		return hcty.False, err
 	}
-	priorC, err := hcty2cty(prior)
-	if err != nil {
-		return hcty.False, err
-	}
-	configC, err := hcty2cty(config)
-	if err != nil {
-		return hcty.False, err
-	}
-	return cty2hcty(objchange.ProposedNew(schema, priorC, configC))
+	priorC := hcty2cty(prior)
+	configC := hcty2cty(config)
+	return cty2hcty(objchange.ProposedNew(schema, priorC, configC)), nil
 }
 
-func htype2ctype(t hcty.Type) (cty.Type, error) {
-	if t.Equals(hcty.NilType) {
-		return cty.NilType, nil
-	}
-	typeJSON, err := hctyjson.MarshalType(t)
-	if err != nil {
-		return cty.Bool, err
+func htype2ctype(t hcty.Type) cty.Type {
+	// Used t.HasDynamicTypes() as an example of how to pattern-match types.
+	switch {
+	case t == hcty.NilType:
+		return cty.NilType
+	case t == hcty.DynamicPseudoType:
+		return cty.DynamicPseudoType
+	case t.IsPrimitiveType():
+		switch {
+		case t.Equals(hcty.Bool):
+			return cty.Bool
+		case t.Equals(hcty.String):
+			return cty.String
+		case t.Equals(hcty.Number):
+			return cty.Number
+		default:
+			contract.Assertf(false, "Match failure on hcty.Type with t.IsPrimitiveType()")
```

Comment:
We have [contract.Failf](https://github.com/pulumi/pulumi/blob/2f3e1657eb37b6c0980765a4e0055e5655b35c0d/sdk/go/common/util/contract/fail.go#L31) for this
```suggestion
			contract.Failf("Match failure on hcty.Type with t.IsPrimitiveType()")
```